### PR TITLE
Make sure DateTime.ToString() works without requiring additional parameters by choosing RFC 3339 as the default.

### PR DESCRIPTION
**Pull request by ahsonkhan**
Enabling `dateTime.ToString()`. Due to method overloading, there are no breaking changes here, and no functionality changes.

Validate and throw that you can't pass fractional format enum when dateformat is rfc1123.

### CODE REVIEW

File: sdk/core/azure-core/src/datetime.cpp [hunk 1]
```diff
@@ -729,42 +729,39 @@ DateTime DateTime::Parse(std::string const& dateTime, DateFormat format)
   }
 }
 
-std::string DateTime::ToString(DateFormat format, TimeFractionFormat fractionFormat) const
+void DateTime::ValidateYear() const
```
Comment (vhvb1989) [inline sdk/core/azure-core/src/datetime.cpp:732]: For Context we use `ThrowIfCancelled()` ,  should we do the same here and change the name to `ThrowIfInvalidDate()` ??
and have one `bool IsValid()`

File: sdk/core/azure-core/src/datetime.cpp [hunk 2]
```diff
@@ -729,42 +729,39 @@ DateTime DateTime::Parse(std::string const& dateTime, DateFormat format)
...
-    if (*this < Year0001)
-    {
-      outOfRange = -1;
-    }
-    else if (*this > Eoy9999)
-    {
-      outOfRange = +1;
-    }
+  static DateTime const Year0001 = DateTime();
+  static DateTime const Eoy9999 = DateTime(9999, 12, 31, 23, 59, 59, 9999999, -1, 0, 0);
```
Comment (vhvb1989) [inline sdk/core/azure-core/src/datetime.cpp:735]: Do we want this as constexpr? in the unnamed namespace?

**Review by vhvb1989 — APPROVED** [review]
Please let @antkmsft review this as the final approver. Or merge if you need and ask Antk to follow

**Review by antkmsft — CHANGES_REQUESTED** [review]
All you need to do to make this change work would be to add default value for the existing argument of the `ToString()` method: `DateFormat format = DateFormat::Rfc3339`. I saw you touched some other things, which were not necessary.

Also, now if I provide DateFormat, I have to provide fraction format. This is bad experience.

Another thing, Jeffrey wanted the DateTimeFormat to always be a required parameter. I am fine if there is a default, but check with him about that.

File: sdk/core/azure-core/src/datetime.cpp (hunk 2 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/src/datetime.cpp:735]: I don't want to modify implementation details in this PR.

File: sdk/core/azure-core/src/datetime.cpp (hunk 1 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/src/datetime.cpp:732]: Those are public APIs, this is not and is just an implementation detail. That said, I do like your suggestion, how about: `ThrowIfUnsupportedYear()`

**Comment by ahsonkhan** [comment]
> All you need to do to make this change work would be to add default value for the existing argument of the `ToString()` method

This wouldn't work because someone passing in `ToString(DateFormat::Rfc1123)` is accidentally passing in some fractional format enum (default value) which is wrong, and meaningless at best.

> I saw you touched some other things, which were not necessary.

That was just moving logic into helper methods to avoid duplication.

> Also, now if I provide DateFormat, I have to provide fraction format. 

How so? There is a new `ToString()` method that just accepts a `DateFormat()`.

@antkmsft please let me know if you have any other **specific** changes that address the primary concern of a simplified and usable `ToString()` helper on DateTime.

**Comment by vhvb1989** [comment]
@JeffreyRichter  please review

**Comment by ahsonkhan** [comment]
> I also added a unit test illustrating the use of a single parameter.

Let's chat about this.

> The issue with this approach is that the caller is now passing in an unnecessary (and potentially invalid) `TimeFractionFormat` value which doesn't apply to Rfc1123. They might now try to pass in different values which won't change the output at all which will be a source of confusion.
> 
> As an aside, there is a general concern that folks don't typically realize optional parameters are optional, and still try to fill them in, which makes the call-site quite cumbersome. Having method overloads simplifies that.

**Review by antkmsft — APPROVED** [review]

**Review by antkmsft — COMMENTED** [review]
Add unit test where it throws if you use fraction format with RFC 11123

**Review by vhvb1989 — APPROVED** [review]

**Pull request merged**